File: system/webservices/sample/loader/csharp/PSLoader.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Web.Services.Protocols;
using System.Xml;
using Loader.RxWebServices;
using RxFaultFactory;

namespace Loader
{
    /// <summary>
    /// This class loads all Content Items defined in a specified data file
    /// (LoaderData.xml) into a specified Rhythmyx Folder. The location of the 
    /// data file and the target Folder are specified in the properties file of the 
    /// loader, Loader.xml.  If the specified target Folder does not exist, it will 
    /// be created. If the Folder contains any Content Items whose sys_title field 
    /// has the same value (case-insensitive) as the sys_title field of a Content 
    /// Item defined in the data file, the existing Content Item will be updated 
    /// with the data from the data file.  If the target Folder does not contain any
    /// Content Items whose sys_title field matches the sys_title field of a Content
    /// Item defined in the data file, a new Content Item will be created.
    /// 
    /// The properties file (Loader.xml) also specifies the Content Type used for the
    /// uploaded Content Items.  The same Content Type is used to upload all Content
    /// Items for a particular run of the loader.
    ///  
    /// The loader returns limited terminal window output while it processes each 
    /// Copntent Item.
    /// </summary>
    class PSLoader
    {
        /// <summary>
        ///     This method is the starting point of the loader program. The parameter
        ///     file, <code>Loader.xml</code>, is assumed to be in the current folder
        ///     of the loader executable program.
        /// </summary>
        /// <param name="args">
        ///     the arguments of the program, which are not used.
        /// </param>
        static void Main(string[] args)
        {
            PSLoader loader = new PSLoader();

            try
            {
                loader.ReadInputData();
                loader.Login();
                loader.CreateTargetFolder();
                loader.UploadItems();
                loader.UploadDataFile();
                loader.Logout();
            }
            catch (SoapException e)
            {
                IPSSoapFault fault = PSFaultFactory.GetFault(e);

                loader.ConsoleMessage("Caught Exception: " + fault.GetFaultName());
                loader.ConsoleMessage("Error Message: " + fault.GetMessage());

                if (fault.GetFaultName() == "PSErrorsFault" ||
                    fault.GetFaultName() == "PSErrorResultsFault")
                {
                    PSErrorsFault errorsFault = (PSErrorsFault)fault;
                    int errorCount = errorsFault.getErrorMessages().Count;
                    int successCount = errorsFault.getSuccessIds().Count;
                    loader.ConsoleMessage("There are " + errorCount + " error messages.");
                    loader.ConsoleMessage("There are " + successCount + " successful ids.");
                }
            }
        }

        /// <summary>
        ///     Updates Content Items read from LoaderData.xml into the target Folder.
        ///     Assumes the target folder already exists in Rhythmyx.
        /// </summary>
        void UploadItems()
        {
            // get the items in the target folder
            PSItemSummary[] curItems = PSWsUtils.FindFolderChildren(m_contService,
                m_props[TARGET_FOLDER]);
      
            foreach (Dictionary<string, string> itemFields in m_itemData)
            {
                PSItemSummary summary = GetItem(curItems, itemFields["sys_title"]);
         
                if (summary != null) // update the existing item
                {
                    UpdateItem(summary.id, itemFields);
                }
                else
                {
                    CreateItem(itemFields);
                }
            }
      
            ConsoleMessage("Finished uploading items into target Folder, " 
                + m_props[TARGET_FOLDER]);
        }

        /// <summary>
        ///     Creates a Content Item of Content Type specified in the properties 
        ///     file (Loader.xml) using the field values defined for the Content Item in 
        ///     the data file (LoaderData.xml).  The created Content Item is added to the
        ///     target Folder specified in the properties file.
        /// </summary>
        /// <param name="fields">
        ///     the set of field values for the created Content Item; assumed not 
        ///     to be <code>null</code> or empty.
        /// </param>
        private void CreateItem(Dictionary<string, string> fields)
        {
            PSItem item = PSWsUtils.CreateItem(m_contService, m_props[CONTENT_TYPE]);
      
            SetItemFields(item, fields);
      
            long id = PSWsUtils.SaveItem(m_contService, item);
            PSWsUtils.CheckinItem(m_contService, id);
            PSWsUtils.TransitionItem(m_sysService, id, "DirecttoPublic");
      
            // Attach the Content Item to the Target folder
            String path = m_props[TARGET_FOLDER];
            PSWsUtils.AddFolderChildren(m_contService, path, new long[]{id});
              
            ConsoleMessage("Created item: " + fields["sys_title"]);
        }

        /// <summary>
        ///     Updates the specified Content Item with the specified field values.
        /// </summary>
        /// <param name="id">
        ///     the ID of the Content Item to update.
        /// </param>
        /// <param name="fields">
        ///     the new values of the specified fields.
        /// </param>
        private void UpdateItem(long id, Dictionary<string, string> fields)
        {
            PSItemStatus status = PSWsUtils.PrepareForEdit(m_contService, id);
      
            PSItem item = PSWsUtils.LoadItem(m_contService, id);
            SetItemFields(item, fields);
            PSWsUtils.SaveItem(m_contService, item);
      
            PSWsUtils.ReleaseFromEdit(m_contService, status);
      
            ConsoleMessage("Updated item: " + fields["sys_title"]);
        }


        /// <summary>
        ///     Sets the fields of the specified Content Item to the specified values.
        /// </summary>
        /// <param name="item">
        ///     the Content Items whose field values are to be updated.
        /// </param>
        /// <param name="fields">
        ///     the set of field values to be updated to the specified 
        ///     Content Item; assumed not to be <code>null</code> or empty.
        /// </param>
        private void SetItemFields(PSItem item, Dictionary<string, string> fields)
        {
           foreach (PSField srcField in item.Fields)
           {
               string nameValue;
               if (fields.TryGetValue(srcField.name, out nameValue))
              {
                 PSFieldValue value = new PSFieldValue();
                 value.RawData = nameValue;
                 srcField.PSFieldValue = new PSFieldValue[] { value };
              }
           }
        }

        /// <summary>
        ///     Determines whether the set of supplied Content Item summaries contains 
        ///     an item with the specified value in the sys_title field.
        /// </summary>
        /// <param name="curItems">
        ///     the set of Content Item summaries to evaluate; assumed 
        ///     not to be  <code>null</code>
        /// </param>
        /// <param name="sysTitle">
        ///     the value of sys_title for which to evaluate; assumed 
        ///     not to be <code>null</code> or empty.
        /// </param>
        /// <returns>
        ///     <code>true</code> if the set of Content Item summaries contains
        ///     a Content Item with the specified value in the sys_title field; 
        ///     <code>false</code> otherwise.
        /// </returns>
        PSItemSummary GetItem(PSItemSummary[] curItems, String sysTitle)
        {
            foreach (PSItemSummary item in curItems)
            {
                if (item.name.ToLower() == sysTitle.ToLower())
                return item;
            }
            return null;

        }

        /// <summary>
        /// Read the input data from the properties file (Loader.xml) and the data 
        /// file (DataFile.xml).
        /// </summary>
        void ReadInputData()
        {
            ConsoleMessage("Read loader parameters...");
            m_props = PSFileUtils.getLoaderProperties();

            ConsoleMessage("Read loader data file, " + m_props[DATA_FILE]
                + " ...");
            m_itemData = PSFileUtils.loadDataFile(m_props[DATA_FILE]);
        }

        /// <summary>
        /// Logs in to the Rhythmyx server using the server connection and
        /// authentication data specified in the loader properties file (Loader.xml)
        /// </summary>
        void Login()
        {
            PSWsUtils.SetConnectionInfo(m_props[PROTOCOL], m_props[HOST], 
                Int16.Parse(m_props[PORT]));

            m_secService = PSWsUtils.GetSecurityService();
            m_rxSession = PSWsUtils.Login(m_secService, m_props[USER_NAME],
                  m_props[PASSWORD], m_props[COMMUNITY], null);

            m_contService = PSWsUtils.GetContentService(m_secService.CookieContainer, 
                m_secService.PSAuthenticationHeaderValue);
            m_sysService = PSWsUtils.GetSystemService(m_secService.CookieContainer, 
                m_secService.PSAuthenticationHeaderValue);
        }

        /// <summary>
        ///     Logs out Rhythmyx for the current Rhythmyx session.
        /// </summary>
        void Logout()
        {
            PSWsUtils.Logout(m_secService, m_rxSession);
        }

        /// <summary>
        ///     Creates any target Folder specified in the properties file (Loader.xml) 
        ///     that does not already exist.
        /// </summary>
        void CreateTargetFolder()
        {
           PSFolder[] folders = PSWsUtils.AddFolderTree(m_contService, 
               m_props[TARGET_FOLDER]);
           
           ConsoleMessage("AddFolderTree operation returned " + folders.Length
                 + " folders");
        }

        ///
        /// <summary>
        ///     Writes the specified message to the console.
        /// </summary>
        /// <param name="msg">
        ///     the console message; assumed not <code>null</code>. 
        /// </param>
        ///
        void ConsoleMessage(String msg)
        {
            Console.WriteLine(msg);
        }

        /// <summary>
        /// Uploads the data file into the target Folder. The data file name
        /// (LoaderData.xml) is specified in the loader properties file (Loader.xml). 
        /// Assumes the target folder already exists in Rhythmyx.
        /// </summary>
        void UploadDataFile()
        {
           string filePath = m_props[DATA_FILE];
           string fileName = getFilename(filePath);

           string targetFolder = m_props[TARGET_FOLDER];
           PSItemSummary[] curItems = PSWsUtils.FindFolderChildren(m_contService,
                 targetFolder);

           PSItemSummary summary = GetItem(curItems, fileName);
          
           if (summary != null) // update the existing item of File Content Type
           {
              UpdateFileItem(summary.id, filePath);
           }
           else
           {
              CreateFileItem(filePath, targetFolder);
           }

           ConsoleMessage("Finished uploading the Data File (" + fileName
              + ") into target Folder, " + targetFolder);
        }

        /// <summary>
        ///     Gets the file name from the specified file path.
        /// </summary>
        /// <param name="filePath">
        ///     the specified file path.
        /// </param>
        /// <returns>
        ///     the file name of the specified file path.
        /// </returns>
        private string getFilename(string filePath)
        {
            int index = filePath.LastIndexOf('\\');
            if (index != -1)
                return filePath.Substring(index + 1);
            else
                return filePath;
        }

        /// <summary>
        ///     Creates a Content Item of File Content Type for the specified file.
        ///     The created Content Item is added to the target Folder.
        /// </summary>
        /// <param name="filePath">
        ///     the file path of the specified file; assumed not to be 
        ///     <code>null</code> or empty.
        /// </param>
        /// <param name="targetFolder">
        ///     the virtual path of the target folder in Rhythmyx; assumed not 
        ///     to be <code>null</code> or empty.
        /// </param>
        void CreateFileItem(string filePath, string targetFolder)
        {
           PSItem item = PSWsUtils.CreateItem(m_contService, "rffFile");

           SetFileItemFields(item, filePath);
              
           long id = PSWsUtils.SaveItem(m_contService, item);
           PSWsUtils.CheckinItem(m_contService, id);
           PSWsUtils.TransitionItem(m_sysService, id, "DirecttoPublic");
             
           // Attach the Content Item to the Target folder
           PSWsUtils.AddFolderChildren(m_contService, targetFolder, new long[] { id });
        }

        /// <summary>
        ///     Updates a Content Item of File Content Type for the specified file.
        /// </summary>
        /// <param name="id">
        ///     the id of the specified Content Item.
        /// </param>
        /// <param name="filePath">
        ///     the file path of the specified file; assumed not to be 
        ///     <code>null</code> or empty.
        /// </param>
        void UpdateFileItem(long id, string filePath) 
        {
           PSItemStatus status = PSWsUtils.PrepareForEdit(m_contService, id);
          
           PSItem item = PSWsUtils.LoadItem(m_contService, id);
          
           // read the orignal binary data before updating
           byte[] content = RetrieveBinaryData(item);
           string text = PSFileUtils.convertBytesToString(content);
           ConsoleMessage("\nThe original content data: \n" + text);
          
           // set the new data
           SetFileItemFields(item, filePath);
           PSWsUtils.SaveItem(m_contService, item);
          
           PSWsUtils.ReleaseFromEdit(m_contService, status);
        }

        /// <summary>
        ///     Retrieves the binary value from the specified Content Item.
        /// </summary>
        /// <param name="item">
        ///     the Content Item from which to retrieve the binary value;
	    ///     assumed it is a Content Item of File Content Type.
        /// </param>
        /// 
        /// <returns>
        ///     the binary value, never <code>null</code>.
        /// </returns>
        byte[] RetrieveBinaryData(PSItem item)
        {
           // Get the value of the binary field, "item"
           PSFieldValue value = null;
           foreach (PSField field in item.Fields)
           {
               if (field.name == "item_file_attachment")
              {
                 PSFieldValue[] values = field.PSFieldValue;
                 value = values[0];
                 break;
              }
           }        

           // Assuming found the "item" field and it is not empty.
           // Retrieves the binary data from the "item" field.
           byte[] binaryData = Convert.FromBase64String(value.RawData);
           return binaryData;
        }

        /// <summary>
        ///     Sets the specified file to the specified File Content Item.
        /// </summary>
        /// <param name="item">
        ///     the File Content Item; assumed not <code>null</code>.
        /// </param>
        /// <param name="filePath">
        ///     the file name of the specified file; assumed not 
        ///     <code>null</code> or empty.
        /// </param>
        void SetFileItemFields(PSItem item, string filePath)
        {
           string fileName = getFilename(filePath);

           foreach (PSField field in item.Fields)
           {
               String name = field.name;
              PSFieldValue newValue = new PSFieldValue();
              if (name == "sys_title" || name == "displaytitle")
              {
                  newValue.RawData = fileName;
                 field.PSFieldValue = new PSFieldValue[] { newValue };
              }
              else if (name == "item_file_attachment")
              {
                 byte[] binaryData = PSFileUtils.ReadBinaryFile(filePath);
                 newValue.RawData = Convert.ToBase64String(binaryData);
                 field.PSFieldValue = new PSFieldValue[] { newValue };
              }
              else if (name == "item_file_attachment_type")
              {
                 newValue.RawData = "text/xml";
                 field.PSFieldValue = new PSFieldValue[] {newValue};
              }
              else if (name == "item_file_attachment_ext")
              {
                 newValue.RawData = ".xml";
                 field.PSFieldValue = new PSFieldValue[] { newValue };
              }
              else if (name == "item_file_attachment_filename")
              {
                  newValue.RawData = fileName;
                  field.PSFieldValue = new PSFieldValue[] { newValue };
              }
              else if (name == "item_file_attachment_size")
              {
                  long size = PSFileUtils.GetFileSize(filePath);
                  newValue.RawData = Convert.ToString(size);
                  field.PSFieldValue = new PSFieldValue[] { newValue };
              }
              else if (name == "sys_workflowid")
              {
                 newValue.RawData = "5";
                 field.PSFieldValue = new PSFieldValue[] { newValue };
              }
           }
        }
   

        /**
         * The Rhythmyx session, initialized by login()}. 
         */
        string m_rxSession;

        /**
         * The security service instance; used to perform operations defined in
         * the security services. It is initialized by login().
         */
        securitySOAP m_secService;

        /**
         * The content service instance; used to perform operations defined in
         * the content services. It is initialized by login().
         */
        contentSOAP m_contService;

        /**
         * The system service instance; used to perform operations defined in
         * the system service. It is initialized by login().
         */
        systemSOAP m_sysService;


        /**
         * The loader properties, read from the file 'Loader.xml'.
         */
        private Dictionary<string, string> m_props;

        /**
         * The Content Item data to be uploaded; read from the file 'DataFile.xml' 
         */
        private List<Dictionary<string, string>> m_itemData;


        /**
         * The property name of the protocol of the server connection.
         */
        public static String PROTOCOL = "Protocol";

        /**
         * The property name of the host of the server connection.
         */
        public static String HOST = "Host";

        /**
         * The property name of the port of the server connection.
         */
        public static String PORT = "Port";

        /**
         * The property name of the name of the login user.
         */
        public static String USER_NAME = "Username";

        /**
         * The property name of the password of the login user.
         */
        public static String PASSWORD = "Password";

        /**
         * The property name of the name of the login Community.
         */
        public static String COMMUNITY = "Community";

        /**
         * The property name of the name of the Content Type of the Content Items to
         * be uploaded. 
         */
        public static String CONTENT_TYPE = "ContentType";

        /**
         * The property name of the target Folder path in Rhythmyx.
         */

        public static String TARGET_FOLDER = "TargetFolder";
        /**
         * The property name of the name of the data file.
         */
        public static String DATA_FILE = "DataFile";
    }
}

File: system/webservices/sample/loader/java/src/com/percussion/webservices/sample/loader/PSLoader.java
/*
 *     Percussion CMS
 *     Copyright (C) 1999-2020 Percussion Software, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     Mailing Address:
 *
 *      Percussion Software, Inc.
 *      PO Box 767
 *      Burlington, MA 01803, USA
 *      +01-781-438-9900
 *      support@percussion.com
 *      https://www.percussion.com
 *
 *     You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>
 */
package com.percussion.webservices.sample.loader;

import com.percussion.webservices.content.ContentSOAPStub;
import com.percussion.webservices.content.PSField;
import com.percussion.webservices.content.PSFieldValue;
import com.percussion.webservices.content.PSFolder;
import com.percussion.webservices.content.PSItem;
import com.percussion.webservices.content.PSItemStatus;
import com.percussion.webservices.content.PSItemSummary;
import com.percussion.webservices.faults.PSContractViolationFault;
import com.percussion.webservices.faults.PSErrorsFault;
import com.percussion.webservices.faults.PSNotAuthenticatedFault;
import com.percussion.webservices.sample.PSWsUtils;
import com.percussion.webservices.security.SecuritySOAPStub;
import com.percussion.webservices.system.SystemSOAPStub;

import java.io.File;
import java.io.InputStream;
import java.net.URL;
import java.rmi.RemoteException;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import javax.activation.DataHandler;
import javax.activation.URLDataSource;
import javax.xml.rpc.ServiceException;

import org.apache.axis.attachments.AttachmentPart;

/**
 * This class loads all Content Items defined in a specified data file
 * (LoaderData.xml) into a specified Rhythmyx Folder. The location of the 
 * data file and the target Folder are specified in the properties file of the
 * loader, Loader.xml.  If the specified target Folder does not exist, it will 
 * be created. If the Folder contains any Content Items whose sys_title field 
 * has the same value (case-insensitive) as the sys_title field of a Content 
 * Item defined in the data file, the existing Content Item will be updated 
 * with the data from the data file.  If the target Folder does not contain any
 * Content Items whose sys_title field matches the sys_title field of a Content
 * Item defined in the data file, a new Content Item will be created.
 *
 * The properties file (Loader.xml) also specifies the Content Type used for the
 * uploaded Content Items.  The same Content Type is used to upload all Content
 * Items for a particular run of the loader.
 * 
 * The loader returns limited terminal window output while it processes each 
 * Content Item.
 */
public class PSLoader
{
   /**
    * This method is the starting point of the loader program. The properties
    * file, <code>Loader.xml</code>, is assumed to be in the same directory as
    * this class.
    *
    * @param args the arguments of the program, which are not used.
    */
   public static void main(String args[])
   {
      PSLoader loader = new PSLoader();

      try
      {
         loader.readInputData();
         loader.login();
         loader.createTargetFolder();
         loader.uploadItems();         
         loader.uploadDataFile();
         loader.logout();
      }
      catch (PSNotAuthenticatedFault e)
      {
         loader.consoleMessage("Caught PSNotAuthenticatedFault: "
               + e.getErrorMessage());
      }
      catch (PSContractViolationFault e)
      {
         loader.consoleMessage("Caught PSContractViolationFault: "
               + e.getErrorMessage());         
      }
      catch (RemoteException e)
      {
         loader.consoleMessage("Caught RemoteException: "
               + e.getMessage());         
      }      
      catch (Exception e)
      {
         e.printStackTrace();
      }
   }
   
   /**
    * Updates Content Items read from LoaderData.xml into the target Folder.
    * Assumes the target folder already exists in Rhythmyx. 
    *
    * @throws Exception if an error occurs.
    */
   void uploadItems() throws Exception
   {
      // get the items in the target folder
      PSItemSummary[] curItems = PSWsUtils.findFolderChildren(m_contService,
            m_props.getProperty(TARGET_FOLDER));
      
      for (Map<String,String> itemFields : m_itemData)
      {
         PSItemSummary summary = getItem(curItems, itemFields.get("sys_title"));
         
         if (summary != null) // update the existing item
         {
            updateItem(summary.getId(), itemFields);
         }
         else
         {
            createItem(itemFields);
         }
      }
      
      consoleMessage("Finished uploading items into target Folder, " 
         + m_props.getProperty(TARGET_FOLDER));
   }

   /**
    * Creates a Content Item of Content Type specified in the properties 
    * file (Loader.xml) using the field values defined for the Content Item in 
    * the data file (LoaderData.xml).  The created Content Item is added to the
    * target Folder specified in the properties file. 
    * 
    * @param fields the set of field values for the created item; assumed not 
    *  to be <code>null</code> or empty.
    */
   private void createItem(Map<String,String> fields) throws Exception
   {
      PSItem item = PSWsUtils.createItem(m_contService, 
         (String)m_props.get(CONTENT_TYPE));
      
      setItemFields(item, fields);
      
      long id = PSWsUtils.saveItem(m_contService, item);
      PSWsUtils.checkinItem(m_contService, id);
      PSWsUtils.transitionItem(m_sysService, id, "DirecttoPublic");
      
      // Attach the Content Item to the Target folder
      String path = m_props.getProperty(TARGET_FOLDER);
      PSWsUtils.addFolderChildren(m_contService, path, new long[]{id});
      
      consoleMessage("Created item: " + fields.get("sys_title"));
   }
   
   /**
    * Updates the specified Content Item with the specified field values.
    * 
    * @param id the ID of the Content Item to update.
    * @param fields the new values of the specified fields.
    *  
    * @throws Exception if an error occurs.
    */
   private void updateItem(long id, Map<String,String> fields)
      throws Exception
   {
      PSItemStatus status = PSWsUtils.prepareForEdit(m_contService, id);
      
      PSItem item = PSWsUtils.loadItem(m_contService, id);
      setItemFields(item, fields);
      PSWsUtils.saveItem(m_contService, item);
      
      PSWsUtils.releaseFromEdit(m_contService, status);
      
      consoleMessage("Updated item: " + fields.get("sys_title"));
   }

   /**
    * Sets the fields of the specified Content Item to the specified values.
    * 
    * @param item the Content Items whose field values are to be updated.
    * @param fields the set of field values to be updated to the specified 
    *    Content Item; assumed not to be <code>null</code> or empty.
    */
   private void setItemFields(PSItem item, Map<String, String> fields)
   {
      for (PSField field : item.getFields())
      {
         if (fields.get(field.getName()) != null)
         {
            PSFieldValue value = new PSFieldValue();
            value.setRawData(fields.get(field.getName()));
            field.setPSFieldValue(new PSFieldValue[]{value});
         }
      }
   }
   
   /**
    * Determines whether the set of supplied Content Item summaries contains 
    * an item with the specified value in the sys_title field.
    * 
    * @param curItems the set of Content Item summaries to evaluate; assumed 
    *    not to be  <code>null</code>
    * @param sysTitle the value of sys_title for which to evaluate; assumed 
    *    not to be <code>null</code> or empty.
    *    
    * @return the summary of the specified item if found, <code>null</code> 
    *    otherwise.
    */
   private PSItemSummary getItem(PSItemSummary[] curItems, String sysTitle)
   {
      for (PSItemSummary item : curItems)
      {
         if (item.getName().equalsIgnoreCase(sysTitle))
            return item;
      }
      return null;
   }
   
   /**
    * Read the input data from the properties file (Loader.xml) and the data 
    * file (DataFile.xml).
    * 
    * @throws Exception if an error occurs.
    */
   void readInputData() throws Exception
   {
      consoleMessage("Read loader parameters...");
      m_props = PSFileUtils.getLoaderProperties();

      consoleMessage("Read loader data file, " + m_props.getProperty(DATA_FILE)
            + " ...");
      m_itemData = PSFileUtils.loadDataFile(m_props.getProperty(DATA_FILE));
   }

   /**
    * Logs in to the Rhythmyx server using the server connection and
    * authentication data specified in the loader properties file (Loader.xml)
    * 
    * @throws PSNotAuthenticatedFault if fails to login or any other error occrus.
    * @throws ServiceException if the method fails to create the new stub.
    */
   void login() throws PSNotAuthenticatedFault, ServiceException
   {
      PSWsUtils.setConnectionInfo(m_props.getProperty(PROTOCOL), 
            m_props.getProperty(HOST), 
            Integer.parseInt(m_props.getProperty(PORT)));
      
      m_secService = PSWsUtils.getSecurityService();
      m_rxSession = PSWsUtils.login(m_secService, 
            m_props.getProperty(USER_NAME),
            m_props.getProperty(PASSWORD),
            m_props.getProperty(COMMUNITY),
            null);      
      
      m_contService = PSWsUtils.getContentService(m_rxSession);
      m_sysService = PSWsUtils.getSystemService(m_rxSession);
   }
   
   /**
    * Logs out Rhythmyx for the current Rhythmyx session.
    *  
    * @throws Exception if any error occurs.
    */
   void logout() throws Exception
   {
      PSWsUtils.logout(m_secService, m_rxSession);
   }
   
   /**
    * Creates any target Folder specified in the properties file (Loader.xml) 
    * that does not already exist.
    * 
    * @throws Exception if an error occurs. 
    */
   void createTargetFolder() throws Exception
   {
      PSFolder[] folders = PSWsUtils.addFolderTree(m_contService, m_props
            .getProperty(TARGET_FOLDER));
      
      consoleMessage("AddFolderTree operation returned " + folders.length
            + " folders");
   }
   
   /**
    * Writes the specified message to the console.
    * 
    * @param msg the console message; assumed not <code>null</code>. 
    */
   void consoleMessage(String msg)
   {
      System.out.println(msg);
   }
   
   
   /**
    * Uploads the data file into the target Folder. The data file name
    * (LoaderData.xml) is specified in the loader properties file (Loader.xml). 
    * Assumes the target folder already exists in Rhythmyx. 
    *
    * @throws Exception if an error occurs.
    */
   void uploadDataFile() throws Exception
   {
      String filename = m_props.getProperty(DATA_FILE);
      String targetFolder = m_props.getProperty(TARGET_FOLDER);
      PSItemSummary[] curItems = PSWsUtils.findFolderChildren(m_contService,
            targetFolder);

      PSItemSummary summary = getItem(curItems, filename);
      
      if (summary != null) // update the existing item of File Content Type
      {
         updateFileItem(summary.getId(), filename);
      }
      else
      {
         createFileItem(filename);
      }
      
      consoleMessage("Finished uploading the Data File (" + filename
         + ") into target Folder, " + targetFolder);
   }
   
   /**
    * Creates a Content Item of File Content Type for the specified file.
    * The created Content Item is added to the target Folder.
    * 
    * @param filename the file name of the specified file; assumed not to be 
    *   <code>null</code> or empty.
    */
   void createFileItem(String filename) throws Exception
   {
      PSItem item = PSWsUtils.createItem(m_contService, "rffFile");
         
      setFileItemFields(item, filename);
         
      long id = PSWsUtils.saveItem(m_contService, item);
      PSWsUtils.checkinItem(m_contService, id);
      PSWsUtils.transitionItem(m_sysService, id, "DirecttoPublic");
         
      // Attach the Content Item to the Target folder
      String path = m_props.getProperty(TARGET_FOLDER);
      PSWsUtils.addFolderChildren(m_contService, path, new long[]{id});
   }

   /**
    * Updates a Content Item of File Content Type for the specified file.
    * 
    * @param id the id of the specified Content Item.
    * @param filename the file name of the specified file; assumed not to be 
    *   <code>null</code> or empty.
    */
   void updateFileItem(long id, String filename) throws Exception
   {
      PSItemStatus status = PSWsUtils.prepareForEdit(m_contService, id);
      
      PSItem item = PSWsUtils.loadItem(m_contService, id);
      
      // read the orignal binary data before update
      byte[] content = retrieveBinaryData(m_contService, item);
      consoleMessage("\nThe original content data: \n" + new String(content));
      
      // set the new data
      setFileItemFields(item, filename);
      PSWsUtils.saveItem(m_contService, item);
      
      PSWsUtils.releaseFromEdit(m_contService, status);
   }
   
   /**
    * Retrieves the binary value from the specified Content Item.
    * 
    * @param binding the stub to which to add the attachement; assumed not
    *    <code>null</code>.
    * @param item the Content Item from which to retrieve the binary value;
    *    assumed it is a Content Item of File Content Type.
    * 
    * @return the binary value, never <code>null</code>.
    */
   byte[] retrieveBinaryData(ContentSOAPStub binding, PSItem item)
      throws Exception
   {
      // Gets the attachment id from the value of the "item" field
      String attachmentId = "";
      for (PSField field : item.getFields())
      {
         if (field.getName().equals("item_file_attachment"))
         {
            PSFieldValue[] values = field.getPSFieldValue();
            attachmentId = values[0].getAttachmentId();
            break;
         }
      }
    
      // Retrieves the attachment with the attachment id
      Object[] attachments = binding.getAttachments();
      for (Object attachment : attachments)
      {
         AttachmentPart part = (AttachmentPart) attachment;
         if (part.getContentId().equals(attachmentId))
         {
            try(InputStream reader = (InputStream) part.getContent()){
               byte[] content = new byte[reader.available()];
               reader.read(content);
               return content;
            }
         }
      }
      return null;

   }

   /**
    * Sets the specified file to the specified File Content Item.
    *  
    * @param item the File Content Item; assumed not <code>null</code>.
    * @param filename the file name of the specified file; assumed not 
    *    <code>null</code> or empty.
    */
   void setFileItemFields(PSItem item, String filename) throws Exception
   {
      URL url = this.getClass().getResource(filename);
      for (PSField field : item.getFields())
      {
         String name = field.getName();
         PSFieldValue newValue = new PSFieldValue();
         if (name.equals("sys_title") || name.equals("displaytitle"))
         {
            newValue.setRawData(filename);               
            field.setPSFieldValue(new PSFieldValue[] { newValue });
         }
         else if (name.equals("item_file_attachment"))
         {
            String attachmentId = addAttachment(m_contService, url);
            newValue.setAttachmentId(attachmentId);
            field.setPSFieldValue(new PSFieldValue[] { newValue });
         }
         else if (name.equals("item_file_attachment_type"))
         {
            newValue.setRawData("text/xml");               
            field.setPSFieldValue(new PSFieldValue[] { newValue });         
         }
         else if (name.equals("item_file_attachment_ext"))
         {
            newValue.setRawData(".xml");               
            field.setPSFieldValue(new PSFieldValue[] { newValue });         
         }
         else if (name.equals("item_file_attachment_filename"))
         {
            newValue.setRawData(filename);               
            field.setPSFieldValue(new PSFieldValue[] { newValue });         
         }
         else if (name.equals("item_file_attachment_size"))
         {
            File file = new File(url.toURI());
            long fileSize = file.length();
            newValue.setRawData(String.valueOf(fileSize));               
            field.setPSFieldValue(new PSFieldValue[] { newValue });         
         }
         else if (name.equals("sys_workflowid"))
         {
            newValue.setRawData("5");
            field.setPSFieldValue(new PSFieldValue[] { newValue });                     
         }
      }
   }
   
   /**
    * Add the supplied attachment to the provided proxy and return the 
    * attachment id.
    * 
    * @param binding the stub to which to add the attachement, assumed not
    *    <code>null</code>.
    * @param attachment the file to attach, assumed not <code>null</code>.
    * @return the attachment id needed on the server to map this to the 
    *    appropriate field, never <code>null</code> or empty.
    */
   private String addAttachment(ContentSOAPStub binding, URL attachment)
   {
      DataHandler handler = new DataHandler(new URLDataSource(attachment));

      AttachmentPart part = new AttachmentPart(handler);
      binding.addAttachment(part);

      return part.getContentId();
   }

   /**
    * The Rhythmyx session, initialized by {@link #login()}. 
    */
   String m_rxSession;
   
   /**
    * The security service instance; used to perform operations defined in
    * the security services. It is initialized by {@link #login()}.
    */
   SecuritySOAPStub m_secService;
   
   /**
    * The content service instance; used to perform operations defined in
    * the content services. It is initialized by {@link #login()}.
    */
   ContentSOAPStub m_contService;
   
   /**
    * The system service instance; used to perform operations defined in
    * the system service. It is initialized by {@link #login()}.
    */
   SystemSOAPStub m_sysService;
   
   /**
    * The loader properties, read from the file 'Loader.xml'.
    */
   private Properties m_props;
   
   /**
    * The Content Item data to be uploaded; read from the file 'DataFile.xml' 
    */
   private List<Map<String, String>> m_itemData;

   /**
    * The property name of the protocol of the server connection. 
    */
   public static String PROTOCOL = "Protocol";

   /**
    * The property name of the host of the server connection. 
    */
   public static String HOST = "Host";

   /**
    * The property name of the port of the server connection.
    */
   public static String PORT = "Port";
   
   /**
    * The property name of the name of the login user.
    */
   public static String USER_NAME = "Username";
   
   /**
    * The property name of the password of the login user.
    */
   public static String PASSWORD = "Password";

   /**
    * The property name of the name of the login Community. 
    */
   public static String COMMUNITY = "Community";
   
   /**
    * The property name of the name of the Content Type of the Content Items to
    * be uploaded. 
    */
   public static String CONTENT_TYPE = "ContentType";
   
   /**
    * The property name of the target Folder path in Rhythmyx. 
    */
   
   public static String TARGET_FOLDER = "TargetFolder";
   
   /**
    * The property name of the name of the data file. 
    */
   public static String DATA_FILE = "DataFile";
}
